nixos/modules/services/web-servers/ttyd.nix

```diff
@@ -0,0 +1,184 @@
+        type = types.nullOr types.str;
+        default = null;
+        example = "eth0";
+        description = "Network interface to bind (eg: eth0)";
+      };
+
+      username = mkOption {
+        type = types.nullOr types.str;
+        default = null;
+        description = "Username for basic authentication.";
+      };
+
+      password = mkOption {
+        type = types.nullOr types.str;
+        default = null;
+        description = ''
+          Password for basic authentication.
+          Warning: this is stored in cleartext in the Nix store!
+         '';
+      };
+
+      signal = mkOption {
+        type = types.ints.u8;
+        default = 1;
+        description = "Signal to send to the command on session close.";
+      };
+
+      clientOptions = mkOption {
+        type = types.attrsOf types.str;
+        default = {};
+        example = literalExample ''{
+          fontSize = "16";
+          fontFamily = "Fira Code";
+
+        }'';
+        description = ''
+          Attribute set of client options for xtermjs.
+          <link xlink:href="https://xtermjs.org/docs/api/terminal/interfaces/iterminaloptions/"/>
+        '';
+      };
+
+      terminalType = mkOption {
+        type = types.str;
+        default = "xterm-256color";
+        description = "Terminal type to report.";
+      };
+
+      checkOrigin = mkOption {
+        type = types.bool;
+        default = false;
+        description = "Whether to allow a websocket connection from a different origin.";
+      };
+
+      maxClients = mkOption {
+        type = types.int;
+        default = 0;
+        description = "Maximum clients to support (0, no limit)";
+      };
+
+      indexFile = mkOption {
+        type = types.nullOr types.path;
+        default = null;
+        description = "Custom index.html path";
+      };
+
+      enableIPv6 = mkOption {
+        type = types.bool;
+        default = false;
+        description = "Whether or not to enable IPv6 support.";
+      };
+
+      enableSSL = mkOption {
+        type = types.bool;
+        default = false;
+        description = "Whether or not to enable SSL (https) support.";
+      };
+
+      certFile = mkOption {
+        type = types.nullOr types.path;
+        default = null;
+        description = "SSL certificate file path.";
+      };
+
+      keyFile = mkOption {
+        type = types.nullOr types.path;
+        default = null;
+        description = "SSL key file path.";
+      };
+
+      caFile = mkOption {
+        type = types.nullOr types.path;
+        default = null;
+        description = "SSL CA file path for client certificate verification.";
+      };
+
+      logLevel = mkOption {
+        type = types.int;
+        default = 7;
+        description = "Set log level.";
+      };
+    };
+  };
+
+  ###### implementation
+
+  config = mkIf cfg.enable {
+
+    warnings = optional (cfg.password != null) "ttyd passwords will be stored as plaintext in the Nix store!";
```

Comment:
Warnings should only be used for things that need to be changed, because they are not silencable. I'd recommend removing the `password` option and instead having a `passwordFile` option, which can be implemented somewhat like this (while removing the `--credentials` flag from `args`):
```nix
{
  systemd.services.ttyd = {
    # ...
    script = if cfg.passwordFile != null then ''
      PASSWORD=$(cat ${escapeShellArg cfg.passwordFile})
      ${pkgs.ttyd}/bin/ttyd ${lib.escapeShellArgs args} \
        --credential ${escapeShellArg cfg.user}:"$PASSWORD" \
        ${pkgs.shadow}/bin/login
    else '' ... '';
  };
}
```

And defining `passwordFile` like
```nix
mkOption {
  type = types.nullOr types.path;
  # This prevents accidentally putting a literal path in the store
  apply = value: if value == null then null else toString value;
  description = ''
    File containing the password to use for basic authentication.
    For insecurely putting the password in the globally readable store use
    <literal>pkgs.writeText "ttydpw" "MyPassword"</literal>.
  '';
}
```

Reply:
Thanks a lot! This was very helpful. Fixed.